src/irmin/append_only_intf.ml

```diff
@@ -26,7 +26,11 @@ module type S = sig
   include Read_only.S
   (** @inline *)
 
-  val add : [> write ] t -> key -> value -> unit Lwt.t
+  type hash
+
+  val index : [> read ] t -> hash -> key option Lwt.t
+
+  val add : [> write ] t -> hash -> value -> key Lwt.t
```

Comment:
How come `Append_only` manipulates hashes and not keys? Is an `Append_only` store always content addressable?

Reply:
`Append_only` is an odd abstraction: it seems to exist entirely in order to be passed to `Content_addressable.Make`.

An earlier version of the patch relied on being able to build a content addressable store from an AO store with hash-like keys, but this isn't currently the case.  I've pushed a commit that reverts this signature change until we actually need it. As a side-effect, this PR no longer touches `src/irmin-{chunk,fs}`, which feels like a win :-)